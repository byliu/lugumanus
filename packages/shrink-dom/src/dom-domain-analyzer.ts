// run in browser

export class DOMDomainAnalyzer {
  private annotatedElements: HTMLElement[] = [];
  private interactiveElements: HTMLElement[] = [];

  private annotationAttribute: string;
  private minInteractiveSize: number;
  private highlightStyle: string;
  private highlightDuration: number;
  private enableHighlight: boolean;

  // 提取默认配置常量
  private static readonly DEFAULT_CONFIG = {
    ANNOTATION_ATTRIBUTE: 'data-element-id',
    MIN_INTERACTIVE_SIZE: 20,
    HIGHLIGHT_STYLE: 'brightness(1.2) contrast(1.1)',
    HIGHLIGHT_DURATION: 500,
    ENABLE_HIGHLIGHT: true,
    TRANSITION_STYLE: 'filter 0.3s ease',
  };

  // 提取交互式标签和属性常量
  static readonly INTERACTIVE_TAGS = ['A', 'BUTTON'];
  static readonly INTERACTIVE_INPUT_TYPES = [
    'button',
    'submit',
    'reset',
    'checkbox',
    'radio',
    'file',
  ];
  static readonly INTERACTIVE_ROLES = ['button', 'link', 'checkbox', 'radio', 'menuitem', 'tab'];
  static readonly NON_EDITABLE_INPUT_TYPES = ['button', 'submit', 'reset', 'hidden', 'file'];

  static readonly ELEMENT_NODE = 1;

  constructor(
    options: {
      annotationAttribute?: string;
      minInteractiveSize?: number;
      highlightStyle?: string;
      highlightDuration?: number;
      enableHighlight?: boolean;
    } = {},
  ) {
    this.annotationAttribute =
      options.annotationAttribute || DOMDomainAnalyzer.DEFAULT_CONFIG.ANNOTATION_ATTRIBUTE;
    this.minInteractiveSize =
      options.minInteractiveSize || DOMDomainAnalyzer.DEFAULT_CONFIG.MIN_INTERACTIVE_SIZE;
    this.highlightStyle =
      options.highlightStyle || DOMDomainAnalyzer.DEFAULT_CONFIG.HIGHLIGHT_STYLE;
    this.highlightDuration =
      options.highlightDuration || DOMDomainAnalyzer.DEFAULT_CONFIG.HIGHLIGHT_DURATION;
    this.enableHighlight =
      options.enableHighlight !== undefined
        ? options.enableHighlight
        : DOMDomainAnalyzer.DEFAULT_CONFIG.ENABLE_HIGHLIGHT;
  }

  public getElements(): HTMLElement[] {
    return [...this.annotatedElements];
  }

  // 抽取检查元素可见性的逻辑
  private isElementVisible(element: HTMLElement, computedStyle: CSSStyleDeclaration): boolean {
    return !(
      computedStyle.display === 'none' ||
      computedStyle.visibility === 'hidden' ||
      computedStyle.opacity === '0' ||
      element.getAttribute('data-visible') === 'false' ||
      element.getAttribute('data-hidden') === 'true' ||
      element.getAttribute('data-ignore') === 'true' ||
      element.getAttribute('hidden') === 'true'
    );
  }

  // 抽取检查元素可交互的逻辑
  private isElementInteractive(element: HTMLElement, computedStyle: CSSStyleDeclaration): boolean {
    const tagName = element.tagName;
    const isInteractiveTag =
      DOMDomainAnalyzer.INTERACTIVE_TAGS.includes(tagName) ||
      (tagName === 'INPUT' &&
        DOMDomainAnalyzer.INTERACTIVE_INPUT_TYPES.includes((element as HTMLInputElement).type)) ||
      (tagName === 'LABEL' && element.hasAttribute('for'));

    const hasInteractiveAttr =
      Boolean(element.onclick) ||
      Boolean(element.onmousedown) ||
      Boolean(element.onmouseup) ||
      Boolean(element.hasAttribute('onclick')) ||
      Boolean(element.hasAttribute('onmousedown')) ||
      Boolean(element.hasAttribute('onmouseup')) ||
      (element.hasAttribute('role') &&
        DOMDomainAnalyzer.INTERACTIVE_ROLES.includes(element.getAttribute('role') || '')) ||
      element.hasAttribute('tabindex');

    return isInteractiveTag || hasInteractiveAttr || computedStyle.cursor === 'pointer';
  }

  // 抽取检查元素可编辑的逻辑
  private isElementEditable(element: HTMLElement): boolean {
    const tagName = element.tagName;
    const isFormElement =
      ((tagName === 'INPUT' &&
        !DOMDomainAnalyzer.NON_EDITABLE_INPUT_TYPES.includes((element as HTMLInputElement).type)) ||
        tagName === 'SELECT' ||
        tagName === 'TEXTAREA') &&
      !(element.hasAttribute('readonly') || element.hasAttribute('disabled'));

    const isContentEditable =
      element.hasAttribute('contenteditable') &&
      element.getAttribute('contenteditable') !== 'false';

    return isFormElement || isContentEditable;
  }

  private traverseNode(node: Node): HTMLElement {
    // 创建元素的克隆副本
    const annotatedNode = node.cloneNode(false) as HTMLElement;

    // 如果节点是元素节点
    if (node.nodeType === DOMDomainAnalyzer.ELEMENT_NODE) {
      const htmlElement = node as HTMLElement;
      const computedStyle = window.getComputedStyle(htmlElement);

      // 将原始元素添加到 annotatedElements 数组中
      this.annotatedElements.push(htmlElement);

      // 获取当前元素在数组中的索引
      const elementIndex = this.annotatedElements.length - 1;
      annotatedNode.setAttribute(this.annotationAttribute, elementIndex.toString());

      // 使用提取的方法简化代码
      const isVisible = this.isElementVisible(htmlElement, computedStyle);
      annotatedNode.setAttribute('data-visible', String(isVisible));

      const isInteractive = this.isElementInteractive(htmlElement, computedStyle);
      annotatedNode.setAttribute('data-interactive', String(isInteractive));

      const isEditable = this.isElementEditable(htmlElement);
      annotatedNode.setAttribute('data-editable', String(isEditable));

      if (isEditable || isInteractive) {
        this.interactiveElements.push(htmlElement);
      }

      // 保存元素的重要属性
      if (htmlElement.hasAttribute('id')) {
        annotatedNode.setAttribute('data-original-id', htmlElement.id);
      }
      if (htmlElement.hasAttribute('class')) {
        annotatedNode.setAttribute('data-original-class', htmlElement.className);
      }
      if (htmlElement.hasAttribute('name')) {
        annotatedNode.setAttribute('data-original-name', htmlElement.getAttribute('name') || '');
      }

      // 对于表单元素，保存其状态
      if (
        htmlElement.tagName === 'INPUT' ||
        htmlElement.tagName === 'TEXTAREA' ||
        htmlElement.tagName === 'SELECT'
      ) {
        const inputElement = htmlElement as
          | HTMLInputElement
          | HTMLTextAreaElement
          | HTMLSelectElement;
        if ('value' in inputElement) {
          annotatedNode.setAttribute('data-value', inputElement.value);
        }
        if ('checked' in inputElement && typeof inputElement.checked === 'boolean') {
          annotatedNode.setAttribute('data-checked', inputElement.checked.toString());
        }
        if ('disabled' in inputElement && typeof inputElement.disabled === 'boolean') {
          annotatedNode.setAttribute('data-disabled', inputElement.disabled.toString());
        }
      }
    }

    for (const child of node.childNodes) {
      const childNode = this.traverseNode(child);
      annotatedNode.appendChild(childNode);
    }

    // 如果是文本节点或其他类型节点，直接克隆
    return annotatedNode;
  }

  private highlightInteractiveElements(): void {
    if (!this.enableHighlight) return;

    // 检查环境是否支持所需的 DOM 功能
    if (!document || !window || typeof document.createElement !== 'function') {
      console.warn('当前环境不支持 DOM 高亮功能');
      return;
    }

    // 过滤出符合条件的交互元素
    const elementsToHighlight = this.interactiveElements.filter((el) => {
      try {
        const computedStyle = window.getComputedStyle(el);
        const width = Number.parseInt(computedStyle.width, 10);
        const height = Number.parseInt(computedStyle.height, 10);
        return width >= this.minInteractiveSize && height >= this.minInteractiveSize;
      } catch (e) {
        console.warn('获取元素样式时出错', e);
        return false;
      }
    });

    // 高亮处理
    const transitionStyle = DOMDomainAnalyzer.DEFAULT_CONFIG.TRANSITION_STYLE;

    for (const el of elementsToHighlight) {
      try {
        // 保存原始样式
        const originalFilter = el.style.filter;

        // 应用高亮样式
        el.style.filter = this.highlightStyle;
        el.style.transition = transitionStyle;

        // 保存原始样式到元素上
        el.setAttribute('data-original-filter', originalFilter);
      } catch (e) {
        console.warn('应用高亮样式时出错', e);
      }
    }

    // 使用setTimeout清理样式
    setTimeout(() => {
      for (const el of elementsToHighlight) {
        try {
          // 恢复原始样式
          el.style.filter = el.getAttribute('data-original-filter') || '';
          el.removeAttribute('data-original-filter');
        } catch (e) {
          console.warn('恢复原始样式时出错', e);
        }
      }
    }, this.highlightDuration);
  }

  public getAnalyzedHTML(element?: HTMLElement): string {
    // 重置内部状态
    this.annotatedElements = [];
    this.interactiveElements = [];

    // 遍历DOM
    const analyzedDocument = this.traverseNode(element ?? document.documentElement);

    // 高亮交互元素
    this.highlightInteractiveElements();

    return analyzedDocument.outerHTML;
  }

  // 静态方法便于直接调用
  public static analyze(options?: {
    annotationAttribute?: string;
    minInteractiveSize?: number;
    highlightStyle?: string;
    highlightDuration?: number;
    enableHighlight?: boolean;
  }): string {
    const analyzer = new DOMDomainAnalyzer(options);
    return analyzer.getAnalyzedHTML();
  }
}
