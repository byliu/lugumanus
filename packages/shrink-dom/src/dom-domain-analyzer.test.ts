import { afterEach, beforeEach, describe, expect, test } from 'bun:test';
import { JSDOM } from 'jsdom';
import { DOMDomainAnalyzer } from './dom-domain-analyzer';

type Global = typeof globalThis & {
  window: Window;
  document: Document;
};

describe('DOMDomainAnalyzer', () => {
  let dom: JSDOM;
  let originalWindow: Global['window'] | undefined;
  let originalDocument: Global['document'] | undefined;

  beforeEach(() => {
    // 保存原始的全局变量
    originalWindow = global.window as Global['window'] | undefined;
    originalDocument = global.document as Global['document'] | undefined;

    // 创建新的 JSDOM 实例
    dom = new JSDOM(`
      <!DOCTYPE html>
      <html>
        <head>
          <title>测试页面</title>
        </head>
        <body>
          <div id="container">
            <button id="btn1">点击我</button>
            <a href="#" id="link1">链接</a>
            <input type="text" id="input1" value="测试输入">
            <input type="checkbox" id="checkbox1">
            <div id="hidden-div" style="display: none;">隐藏元素</div>
            <div id="pointer-div" style="cursor: pointer;">可点击元素</div>
            <div id="regular-div">普通元素</div>
            <label for="input1">输入标签</label>
            <textarea id="textarea1">文本区域</textarea>
            <div id="editable-div" contenteditable="true">可编辑内容</div>
          </div>
        </body>
      </html>
    `);

    // 设置全局变量
    global.window = dom.window as unknown as Window & typeof globalThis;
    global.document = dom.window.document;

    // 模拟 window.getComputedStyle
    global.window.getComputedStyle = (element: Element) => {
      const style = {} as CSSStyleDeclaration;

      // 为测试元素设置默认样式
      style.display = 'block';
      style.visibility = 'visible';
      style.opacity = '1';
      style.width = '100';
      style.height = '50';
      style.cursor = 'default';

      // 为特定元素设置特殊样式
      if (element.id === 'hidden-div') {
        style.display = 'none';
      }

      if (element.id === 'pointer-div') {
        style.cursor = 'pointer';
      }

      return style;
    };

    // 确保 Element.prototype 上有 style 属性的所有需要的方法
    if (typeof Element !== 'undefined' && Element.prototype) {
      // 扩展 Element.prototype.style 以支持所有需要的属性
      Object.defineProperty(Element.prototype, 'style', {
        get: function () {
          if (!this._style) {
            this._style = {
              filter: '',
              transition: '',
              setProperty: function (prop: string, value: string) {
                this[prop] = value;
              },
              getPropertyValue: function (prop: string) {
                return this[prop] || '';
              },
            };
          }
          return this._style;
        },
        configurable: true,
      });
    }
  });

  afterEach(() => {
    // 恢复原始全局变量
    global.window = originalWindow as Global['window'];
    global.document = originalDocument as Global['document'];
  });

  test('getAnalyzedHTML 应该返回带注释的 HTML', () => {
    // 创建分析器实例
    const analyzer = new DOMDomainAnalyzer();

    // 获取分析后的 HTML
    const analyzedHTML = analyzer.getAnalyzedHTML();

    // 验证基本结构
    expect(analyzedHTML).toContain('<html');

    // 验证元素被正确注释
    expect(analyzedHTML).toContain(`data-interactive="true"`);
    expect(analyzedHTML).toContain(`data-editable="true"`);
    expect(analyzedHTML).toContain(`data-visible="true"`);
    expect(analyzedHTML).toContain(`data-original-id="`);
  });

  test('静态方法 analyze 应该返回带注释的 HTML', () => {
    // 直接调用静态方法
    const analyzedHTML = DOMDomainAnalyzer.analyze();

    // 验证基本结构
    expect(analyzedHTML).toContain('<html');

    // 验证元素被正确注释
    expect(analyzedHTML).toContain(`data-interactive="true"`);
    expect(analyzedHTML).toContain(`data-editable="true"`);
    expect(analyzedHTML).toContain(`data-visible="true"`);
    expect(analyzedHTML).toContain(`data-original-id="`);
  });

  test('可以通过选项自定义注释属性', () => {
    // 创建使用自定义注释属性的分析器
    const customAttributeName = 'data-custom-id';
    const analyzer = new DOMDomainAnalyzer({
      annotationAttribute: customAttributeName,
    });

    // 获取分析后的 HTML
    const analyzedHTML = analyzer.getAnalyzedHTML();
    const dom = new JSDOM(analyzedHTML);
    const document = dom.window.document;

    // 验证自定义属性被设置在文档元素上
    const button = document.getElementById('btn1');
    expect(button).not.toBeNull();
    expect(button?.hasAttribute(customAttributeName)).toBe(true);
  });

  test('互动元素应被正确识别', () => {
    // 创建分析器
    const analyzer = new DOMDomainAnalyzer();

    // 获取分析后的 HTML
    const analyzedHTML = analyzer.getAnalyzedHTML();
    const dom = new JSDOM(analyzedHTML);
    const document = dom.window.document;

    // 检查按钮元素
    const analyzedButton = document.querySelector('[data-element-id]#btn1');
    expect(analyzedButton).not.toBeNull();

    // 检查链接元素
    const analyzedLink = document.querySelector('[data-element-id]#link1');
    expect(analyzedLink).not.toBeNull();

    // 检查输入框
    const analyzedInput = document.querySelector('[data-element-id]#input1');
    expect(analyzedInput).not.toBeNull();

    // 检查带有指针样式的 div
    const pointerDiv = document.querySelector('[data-element-id]#pointer-div');
    expect(pointerDiv).not.toBeNull();
  });

  test('getAnalyzedHTML 输出应符合快照', () => {
    // 创建一个具有确定性配置的分析器
    const analyzer = new DOMDomainAnalyzer({
      enableHighlight: false, // 禁用高亮以避免异步行为
    });

    // 获取分析后的 HTML
    const analyzedHTML = analyzer.getAnalyzedHTML();

    // 移除可能因平台差异而变化的空白符
    const normalizedHTML = analyzedHTML.replace(/\s+/g, ' ').trim();

    // 使用快照测试验证输出
    expect(normalizedHTML).toMatchSnapshot();
  });

  test('高亮功能应该在测试环境中不会抛出错误', () => {
    // 创建分析器，启用高亮
    const analyzer = new DOMDomainAnalyzer({
      enableHighlight: true,
      highlightDuration: 0, // 使用 0 毫秒确保同步执行
    });

    // 应该能顺利执行而不抛出错误
    expect(() => {
      analyzer.getAnalyzedHTML();
    }).not.toThrow();
  });

  test('分析器应该正确标记可编辑元素的快照', () => {
    // 创建分析器实例
    const analyzer = new DOMDomainAnalyzer({
      enableHighlight: false, // 禁用高亮以确保一致性
    });

    const analyzedHTML = analyzer.getAnalyzedHTML();
    expect(analyzedHTML).toMatchSnapshot();
  });
});
