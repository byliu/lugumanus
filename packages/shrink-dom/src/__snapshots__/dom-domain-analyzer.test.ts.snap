// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`DOMDomainAnalyzer getAnalyzedHTML 输出应符合快照 1`] = `"<html data-element-id="0" data-visible="true" data-interactive="false" data-editable="false"><head data-element-id="1" data-visible="true" data-interactive="false" data-editable="false"> <title data-element-id="2" data-visible="true" data-interactive="false" data-editable="false">测试页面</title> </head> <body data-element-id="3" data-visible="true" data-interactive="false" data-editable="false"> <div id="container" data-element-id="4" data-visible="true" data-interactive="false" data-editable="false" data-original-id="container"> <button id="btn1" data-element-id="5" data-visible="true" data-interactive="true" data-editable="false" data-original-id="btn1">点击我</button> <a href="#" id="link1" data-element-id="6" data-visible="true" data-interactive="true" data-editable="false" data-original-id="link1">链接</a> <input type="text" id="input1" value="测试输入" data-element-id="7" data-visible="true" data-interactive="false" data-editable="true" data-original-id="input1" data-value="测试输入" data-checked="false" data-disabled="false"> <input type="checkbox" id="checkbox1" data-element-id="8" data-visible="true" data-interactive="true" data-editable="true" data-original-id="checkbox1" data-value="on" data-checked="false" data-disabled="false"> <div id="hidden-div" style="display: none;" data-element-id="9" data-visible="false" data-interactive="false" data-editable="false" data-original-id="hidden-div">隐藏元素</div> <div id="pointer-div" style="cursor: pointer;" data-element-id="10" data-visible="true" data-interactive="true" data-editable="false" data-original-id="pointer-div">可点击元素</div> <div id="regular-div" data-element-id="11" data-visible="true" data-interactive="false" data-editable="false" data-original-id="regular-div">普通元素</div> <label for="input1" data-element-id="12" data-visible="true" data-interactive="true" data-editable="false">输入标签</label> <textarea id="textarea1" data-element-id="13" data-visible="true" data-interactive="false" data-editable="true" data-original-id="textarea1" data-value="文本区域" data-disabled="false">文本区域</textarea> <div id="editable-div" contenteditable="true" data-element-id="14" data-visible="true" data-interactive="false" data-editable="true" data-original-id="editable-div">可编辑内容</div> </div> </body></html>"`;

exports[`DOMDomainAnalyzer 分析器应该正确标记可编辑元素的快照 1`] = `
"<html data-element-id="0" data-visible="true" data-interactive="false" data-editable="false"><head data-element-id="1" data-visible="true" data-interactive="false" data-editable="false">
          <title data-element-id="2" data-visible="true" data-interactive="false" data-editable="false">测试页面</title>
        </head>
        <body data-element-id="3" data-visible="true" data-interactive="false" data-editable="false">
          <div id="container" data-element-id="4" data-visible="true" data-interactive="false" data-editable="false" data-original-id="container">
            <button id="btn1" data-element-id="5" data-visible="true" data-interactive="true" data-editable="false" data-original-id="btn1">点击我</button>
            <a href="#" id="link1" data-element-id="6" data-visible="true" data-interactive="true" data-editable="false" data-original-id="link1">链接</a>
            <input type="text" id="input1" value="测试输入" data-element-id="7" data-visible="true" data-interactive="false" data-editable="true" data-original-id="input1" data-value="测试输入" data-checked="false" data-disabled="false">
            <input type="checkbox" id="checkbox1" data-element-id="8" data-visible="true" data-interactive="true" data-editable="true" data-original-id="checkbox1" data-value="on" data-checked="false" data-disabled="false">
            <div id="hidden-div" style="display: none;" data-element-id="9" data-visible="false" data-interactive="false" data-editable="false" data-original-id="hidden-div">隐藏元素</div>
            <div id="pointer-div" style="cursor: pointer;" data-element-id="10" data-visible="true" data-interactive="true" data-editable="false" data-original-id="pointer-div">可点击元素</div>
            <div id="regular-div" data-element-id="11" data-visible="true" data-interactive="false" data-editable="false" data-original-id="regular-div">普通元素</div>
            <label for="input1" data-element-id="12" data-visible="true" data-interactive="true" data-editable="false">输入标签</label>
            <textarea id="textarea1" data-element-id="13" data-visible="true" data-interactive="false" data-editable="true" data-original-id="textarea1" data-value="文本区域" data-disabled="false">文本区域</textarea>
            <div id="editable-div" contenteditable="true" data-element-id="14" data-visible="true" data-interactive="false" data-editable="true" data-original-id="editable-div">可编辑内容</div>
          </div>
        
      
    </body></html>"
`;
